### ISSUE TITLE
`m.redraw()` and `m.redraw.sync()` synthesis

### ISSUE THREAD

**Issue by pygy** [MithrilJS/mithril.js#1765]
Building on top of the #1166, #1592, #1728 and #1751 discussions

### Monolithic render. All roots are rendered in sequence. #1751, @isiahmeadows

```JS
var roots = []
var callbacks = []
function renderRoots() {
  callbacks.forEach(c=>c())
}
```

### Trigger an extra redraw when done #1728 @isiahmeadows

```JS
var rendering = false
var again = false
function sync() {
    if (rendering) {
        again = true
    } else {
        do {
            again = false
            rendering = true
            renderRoots()
            rendering = false
        } while (again)
    }
}
```

### Async `m.redraw()`, `m.redraw.sync()` #1592

```JS
// 
function throttle(callback) {
	//60fps translates to 16.6ms, round it down since setTimeout requires int
	var delay = 16
	var last = 0, pending = null
	var timeout = typeof requestAnimationFrame === "function" ? requestAnimationFrame : setTimeout
	return function() {
		var elapsed = Date.now() - last
		if (pending === null) {
			pending = timeout(function() {
				pending = null
				callback()
				last = Date.now()
			}, delay - elapsed)
		}
	}
}

var redraw = throttle(sync) // leaving the possibility to inject a custom `throttle` for the tests
redraw.sync = sync
```

### Wait for the end of the bubbling phase before triggering a sync redraw?

We may also want to ensure that `sync` redraws triggered in events wait for the end of the bubble phase (in order to avoid #1751-like issues), but that would probably require, if not a synthetic event system, active collaboration from `render`... 

We may add support for `e.redraw = 'sync'` that would set a flag somewhere.

`addEvent` could add a corresponding event on the root, which would trigger the sync redraw if the flag is set.

```JS
// in api/redraw.js
var syncScheduled = false
renderService.setEventCallback(function(e) {
  if  (e.redraw === 'sync') syncScheduled = true
  else if (e.redraw !== false) redraw()
})

renderService.setRootEvents(function(reset) {
  if (reset) syncScheduled = false // reset on render() start in case the previous one threw
  else if  (syncScheduled === true) {
    syncScheduled = false
    m.redraw.sync()
  }
})

// in render/render.js
function updateEvent(vnode, value, key) {
  // ...
  if (typeof rootEvent === 'function') root.key = rootEvent
  // ...
}
```

That could still be problematic if a user action triggers more than one event (`onmousedown` => `onclick`) and a redraw is triggered before the last one fires. Suppose that `onmousedown` triggers a sync redraw, and the update adds an `onclick` handler somewhere in the same branch of the DOM tree... I don't think we could prevent these from firing. So this last proposal makes sync redraws from events safer, but it's not bulletproof...

So I'd keep async, throttled redraws as a default, and leave the `sync` ones for the special cases, with appropriate *caveat emptor*s in the docs.

---

Mounting (routed or not) should only cause a render of the corresponding root (sync unless delayed by a RouteReslover), that is what #1592 does.

---

@lhorie @tivac @isiahmeadows I propose we turn #1592 into its own branch in the repo to make it easier to collaborate on this (pushing to a PR branch can lead to conflicts and is rather rigid). Alternatively, we can merge #1592  and add the rest afterwards.

**Comment by pygy** [comment]
How would that behave in practice: bar bugs in the design:

- `m.redraw()` is always throttled, async, meaning it can be called from anywhere, it will launch the `render()` calls from a clean stack at max 60 FPS.
- `m.redraw.sync()` can be called safely:
  - from non-view code
  - from hooks
  - from event handlers as long as a later event handler isn't added uptree on update. e.g. `onclick` fires later than `onmousedown` for the same user action.

**Comment by isiahmeadows** [comment]
@pygy 

I'll start with the meta here:

> I propose we turn #1592 into its own branch in the repo to make it easier to collaborate on this (pushing to a PR branch can lead to conflicts and is rather rigid). Alternatively, we can merge #1592 and add the rest afterwards.

I'm not sure we need to create a *full* fork/branch just for this discussion, because it's only really going to affect 1 file (`api/redraw.js`) for the most part, maybe small parts of 2-3 others (sans tests). I only created the gist as a PoC for an idea, but even that has issues. Writing code isn't really going to help us much because the semantics are straightforward for any of us to implement, and would only take 1-2 days for any of us to put together a PR for the right semantics.

I do feel there are things worth mentioning:

1. Clearly, the current behavior is ill-defined and confusing, and nobody is arguing it.
1. Leo hasn't been as involved in the current discussion as he has historically been.
1. We've been discussing this intermittently for almost a year now - #1166, stemming from [#1100 (comment)](https://github.com/lhorie/mithril.js/issues/1100#issuecomment-233863732). That's also when the `m.redraw.sync()` idea was presented, and we shouldn't forget that lengthy pre-1.0 discussion.
1. Making `m.redraw()` always async will be a breaking change, since many people already expect a synchronous redraw in certain conditions, and it's currently documented to be sometimes-sync (even though [the docs are only half-right](https://github.com/lhorie/mithril.js/issues/1751)).

-----

Now for the actual subject matter itself:

- If we make `m.redraw()` purely async, we could easily use the monolithic render idea I came up with.
- `m.redraw.sync()` is compatible with the monolithic redraw, provided `m.redraw()` is made 100% async.
- I'd consider deferring sync redraws until after the bubbling phase more of a bug fix, and you could get around the issue of spammy events like `onmousemove` by invoking a throttled, always-async redraw. Much easier and less magical than creating a synthetic event system like React's.
- Yes, making `m.redraw()` purely async would also help with responsiveness, since we will own the async redraw loop and will be able to time ourselves accordingly. React has taken advantage of that control to [experiment with incremental, continuous rendering via specialized cancellable continuations](https://github.com/acdlite/react-fiber-architecture), stuff you don't normally see outside of game engines and other projects with similarly high performance demands.
  - I question why React even needs that, though - if they'd simplify their internal state model, they'd see some immediate improvements in memory, speed, and bundle size, arguably larger than what they'd initially achieve with their Fibers experiment. I mean, I'm 0% surprised that [Preact achieved 90% of the primary functionality](https://github.com/developit/preact/wiki/Differences-to-React) in literally ~10% of the code size.

> from event handlers as long as a later event handler isn't added uptree on update. e.g. `onclick` fires later than `onmousedown` for the same user action.

Under what circumstances would this be theoretically unsafe (and not just a potential perf issue)?

**Comment by pygy** [comment]
@isiahmeadows 

Meta: 

- (0) to branch or not branch: the code changes are trivial (a little more than the samples above), there's much more work on the tests side of things though. I can work on this, but I you want to do it as well, you should know that I'm used to squash my own branches, which will make collaboration tricky (i.e. expect lost commits as I force-push, those are automatisms).

- (1) yup and (2) that's among other things the reason why I opened this, to summarize the topic and propose a path forward, with a coherent vision.

- (3) It actually started with #928. A lot has been said indeed. I was initially against monolithic redraws because I kind of liked the unofficial `root.redraw()` methods, but they're gone now...

- (4) The docs also say "treat it as if it were always async" (which isn't possible in possible right now e.g. in the hooks/event handlers, `setTimeout(m.redraw())` is needed there).

----

Meat:

So basically you agree with the proposal :-)

Yes, the throttled redraws would be built on top of the monolithic, "one again" system you proposed. It makes them compatible/composable (you could `m.redraw.sync()` from the hooks of a `m.redraw()`-triggered render).

> Under what circumstances would this be theoretically unsafe (and not just a potential perf issue)?

#1751 redux.

http://jsbin.com/qahusomuwa/edit?html,console,output

The `onmousedown` handler causes a sync "redraw" after the end of the bubbling phase. That redraw adds an onclick handler on the root, which fires.

This can be problematic if you use more than one event handler for cross-platform compat (`ontouchstart`, `onclick`, ...).

**Comment by isiahmeadows** [comment]
@pygy 

> So basically you agree with the proposal :-)

In a very long-winded way, yes. 😄 

> > Under what circumstances would this be theoretically unsafe (and not just a potential perf issue)?
>
> [reply...]

**Comment by isiahmeadows** [comment]
I'm just glad we're finally approaching a reasonable, sane conclusion after all this time! 🎉

**Comment by pygy** [comment]
:Tada: :-)

Re. tests, #1592 already ported most tests to a pseudo-async system, but I'd rather duplicate them to have instances of `redraw()` and `redraw.sync()` for each... We'll also have to add tests for triggering sync and async redraws from the hooks and from events, add tests for the monolithic nature of things, and maybe router-specific tests as well (not sure)...

**Comment by isiahmeadows** [comment]
#1592 will need partially redone anyways to go with our idea of a top-level monolithic redraw rather than a per-root throttle. It may affect timings, too.

**Comment by pygy** [comment]
I was hoping to build on top of it, actually...

It has the right `throttle` implementation, and I'd rather not redo the tests changes...

We'd have to change
```JS
	function subscribe(key, callback) {
		unsubscribe(key)
		callbacks.push(key, _throttle(callback))
	}
	// ...
	function redraw() {
		for (var i = 1; i < callbacks.length; i += 2) {
			callbacks[i]()
		}
	}
```

into 
```JS
	function subscribe(key, callback) {
		unsubscribe(key)
		callbacks.push(key, callback)
	}
	// ...
	function runAll() {
		for (var i = 1; i < callbacks.length; i += 2) {
			callbacks[i]()
		}
	}
	var redraw = throttle(runAll)
```

To get monolithic, async redraws.

Then we can implement `sync()` on top of `runAll()` and define `redraw = throttle(sync)`.

**Comment by isiahmeadows** [comment]
Okay. I see what you mean.

**Comment by pygy** [comment]
One last thing about this: we can also throw an error if `render` is called from hooks on a root that's already being rendered to (mark the root.rendering = true, wrap the `render()` body in a try/finally block and unlock in finally).

**Comment by isiahmeadows** [comment]
@pygy We could also just stuff the currently rendering roots into an array stack, and `indexOf` test it before attempting to render, so we aren't needlessly adding a property to an object. The common case is literally one item in it, so it'll make near-zero difference performance-wise.

**Comment by tivac** [comment]
Based on our current semver policy this would be part of `1.2.0` because we're afraid of major-version bumps.

I'm fine with these changes, though the intricacies of some of the details are beyond me currently (and maybe ever). I've never used multiple mountpoints so I have a hard time generating an opinion there.

My forever-concern with any large change like this will be perf/complexity, naturally.

Oh, and one minor code thing,

```js
var timeout = typeof requestAnimationFrame === "function" ? requestAnimationFrame : setTimeout
```

this check should only be done once though, `rAF` isn't likely to appear later on in the page's lifecycle. 😄

**Comment by pygy** [comment]
It will affect perf in that default redraws will be delayed by an average of 8ms ([0-16], uniformly distributed). For the average user, UI changes that happen within 100ms are perceived as instantaneous, so it may push some borderline updates from instant to near-instant.

The major win is on the complexity/reliability. Currently, `onclick: m.route('/x')` will redraw twice, once synchronoulsy on click and one 16ms later because of the router. Double redraws can also be triggered by user code.

After this change, users will be able to spam m.redraw() calls and expect predictable behavior. Likewise, calling `redraw.sync()` would be safe from lifecycle events, even in multi-mount point apps. `redraw.sync();redraw.sync()` called from the hooks of a single `render` would only re-render once.

Regarding `timeout` you're right but `throttle` will only be called once anyway with the new plan. With #1592 as is it is once per `mount()` call, so in that scenario it would have been better moved out, if for minute gains.

**Comment by tivac** [comment]
I'm less worried about redraw timing (due to perception limits you quoted), more worried about actual code perf. Redraw timing code isn't likely to be trouble though, but I still want a way to verify all our perf assumptions.

**Comment by pygy** [comment]
This doesn't touch `m.render()` only the time at which it is called, so I'd expect it to be perf-neutral. 

Another thing I had forgotten is that it opens a new possibility to shoot yourself in the foot with `onupdate: m.redraw.sync` => infinite, sync redraw loop; please close your tab/browser.

The pool work I've also started could have a perf impact (possibly positive), as would a #1773 fix (in either direction). It would be nice to have a benchmark system in place to evaluate them.

**Comment by isiahmeadows** [comment]
@tivac The redraw timing is outside the critical path, and would likely only get baseline optimizations. In practice, I'd be surprised if it took >1ms in its entirety, compared to ~10-15ms for a full render. And also, the my idea of a monolithic redraw would actually *simplify* the code paths at runtime, since it's not having to call across closures to do it. [Here's a link to the gist I originally thought this out in](https://gist.github.com/isiahmeadows/9898852646f4b5e10f403a22e82c88cb), although I've since edited it to incorporate some of the newer ideas that have come up since.

(@pygy I've edited that gist down some and to incorporate an async-only redraw. It's actually now smaller than the current `api/redraw.js` by a few lines.)

**Comment by isiahmeadows** [comment]
Note that my gist is untested, so I may have missed something. But it likely should work, since it's nearly trivial now.

**Comment by isiahmeadows** [comment]
@pygy We can dodge the infinite loop issue with `onupdate: m.redraw.sync` by throwing an error in `m.render` when requesting a render from a node while currently rendering to that node.

**Comment by pygy** [comment]
@isiahmeadows I'm talking about nested `m.redraw.sync()` here, not `m.render()`. As described above, a request for the former waits for the end of the current monolithic render before starting the next one... so that's a footgun. We could instead cap the number of consecutive sync redraws as a safety measure (such loops are bugs, always).

**Comment by pygy** [comment]
@isiahmeadows also, the `elapsed` logic in https://github.com/pygy/mithril.js/blob/cb44c556533a70e393abf2c05b79dbc878e32b3b/api/redraw.js is  necessary to get 60 FPS in IE9, and We need to make `throttle` swappable for the tests, so in its current shape, the gist is too simple.

**Comment by isiahmeadows** [comment]
@pygy

> I'm talking about nested m.redraw.sync() here, not m.render()

`m.redraw.sync()` will likely call `m.render()` directly, just like what `m.redraw()` already does now. But we can do the logic there instead (added to my gist).

> also, the elapsed logic in [...]

I've fixed my gist to incorporate that. It was a pretty simple fix. I also reverted most of the other more superficial changes.

**Comment by pygy** [comment]
@isiahmeadows Now you've lost the "one again" nature of `m.redraw.sync()` that you originally proposed, and that is IMO good, and part of OP... You can't even call `redrawSync()` from a hook of a normal `redraw()` with the code as is.

`onupdate: m.redraw.sync` is a footgun and we may want to cap the number of times it can be "chained" but zero is too low. One or two would be better IMO.

Locking nested `render()` calls IMO belongs to `render` itself.

**Comment by isiahmeadows** [comment]
@pygy 

> Now you've lost the "one again" nature of m.redraw.sync() that you originally proposed, and that is IMO good

Yeah...I quickly abandoned it for obvious reasons, falling back to the always-async or always-sync idea.

> `onupdate: m.redraw.sync` is a footgun and we may want to cap the number of times it can be "chained" but zero is too low. One or two would be better IMO.

You mean "nested", not "chained", right? I do currently disallow nesting, but note that `onupdate: m.redraw.sync` is already unnecessary because `m.redraw()` redraws every mounted root to begin with. You only need it if you're using a different `api/redraw` instance than the global one, and in that case, it won't conflict.

> Locking nested `render()` calls IMO belongs to `render` itself.

Which is why I initially proposed putting the lock there. The safety issues are with nested `render()` calls, not nested `redraw()` calls.

**Comment by pygy** [comment]
This was merged in `next` a while ago.
